### PR TITLE
Slingshot Editorial

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

lmk if the explanation should be a bit more thorough on the case handling and other stuff (altho I think writing out all the casework is kind of a waste of time)

### CODE REVIEW

File: solutions/platinum/usaco-816.mdx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+template <class T> class MinSegmentTree {
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:35]: codesnip

File: solutions/platinum/usaco-816.mdx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+				if (pref.get(b_id) > add_1) { pref.set(b_id, add_1); }
+
+				// applying this result for future queries where b >= y.
+				ll add_2 = (t) ? c - a + b : c + a + b;
+				if (suff.get(b_id) > add_2) { suff.set(b_id, add_2); }
+			} else {
+				// res_1 is for if we are choosing a slingshot with b <= y
+				// res_2 is for if we are choosing a slingshot with b >= y
+				ll res_1 = min(INF, pref.range_min(0, b_id + 1));
+				ll res_2 = min(INF, suff.range_min(b_id, coords.size()));
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:131]: honestly just `res1` and `res2` is ok

File: solutions/platinum/usaco-816.mdx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	// apply coordinate compression on all relevant locations
+	sort(begin(coords), end(coords));
+	coords.erase(unique(begin(coords), end(coords)), end(coords));
+	const auto id = [&](int x) -> int {
+		return lower_bound(begin(coords), end(coords), x) - begin(coords);
+	};
+
+	constexpr ll INF = 1e18;
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:95]: int max?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/platinum/usaco-816.mdx (hunk 1 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx:35]: oops

File: solutions/platinum/usaco-816.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx:95]: $t + a + b$ can be greater than int overflow so I don't think that would be ideal

I can probably make it smth smaller tho

Comment (SansPapyrus683) [reply solutions/platinum/usaco-816.mdx:97]: no it's fine just add a comment "1e18 bc of overflow..."
then where it overflows "right here"

**Comment by SansPapyrus683** [comment]
you say 4 cases but i don't see any separation of them in the code

**Comment by TheGamingMousse** [comment]
> you say 4 cases but i don't see any separation of them in the code

I mentioned that I didn't code up each of the four cases separately in the editorial.

Basically, the loop considers if $a_i <= x_i$ for every matching, and then the $pref$ and $res1$ thing matches up with if $b_i <=  y_i$ or something, and similarly. lmk if I should make this a bit more clear.

File: solutions/platinum/usaco-816.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx]: done

File: solutions/platinum/usaco-816.mdx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+				if (pref.get(b_id) > add_1) { pref.set(b_id, add_1); }
+
+				// applying this result for future queries where b >= y.
+				ll add_2 = (t) ? c - a + b : c + a + b;
+				if (suff.get(b_id) > add_2) { suff.set(b_id, add_2); }
+			} else {
+				// res1 is for if we are choosing a slingshot with b <= y
+				// res2 is for if we are choosing a slingshot with b >= y
+				// we set res1 and res2 to be at most INF to prevent overflow when
+				// adding to them
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:133]: multiline comment

File: solutions/platinum/usaco-816.mdx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+				ll res1 = min(INF, pref.range_min(0, b_id + 1));
+				ll res2 = min(INF, suff.range_min(b_id, coords.size()));
+
+				if (t == 0) {
+					res1 += b - a;
+					res2 -= a + b;
+				} else {
+					res1 += a + b;
+					res2 += a - b;
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:142]: can i has an indicator "possible overflow here"

File: solutions/platinum/usaco-816.mdx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		     [&](const array<int, 4> &x, const array<int, 4> &y) -> bool {
+			     int v_1 = x[1], v_2 = y[1];
+			     if (t == 0) { v_1 *= -1, v_2 *= -1; }
+			     return (v_1 == v_2) ? x[0] < y[0] : v_1 < v_2;
+		     });
+
+		/*
+		 * pref = RMQ for best option if b <= y
+		 * suff = RMQ for best option if b >= y
+		 */
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-816.mdx:116]: this can actually just be two single lines but you can keep as is

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/platinum/usaco-816.mdx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+				// could have overflow when adding (e.g. if adding a + b to res1)
+				if (t == 0) {
+					res1 += b - a;
+					res2 -= a + b;
+				} else {
+					res1 += a + b;
+					res2 += a - b;
+				}
+
+				res[c] = min({res[c], res1, res2, (ll)abs(a - b)});
```
Comment (ryanchou-dev) [inline solutions/platinum/usaco-816.mdx:146]: wait isn't `llabs` its own function

File: solutions/platinum/usaco-816.mdx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+id: usaco-816
+source: USACO Platinum February 2018
+title: Slingshot
+author: Justin Ji
+---
+
+## Explanation
+
+If we need to move from $a$ to $b$ with a slingshot from $x$ to $y$ that takes
+$t$ time, then it will take $abs(a - x) + abs(b - y) + t$ time.
```
[the remark this answers is not in the document]
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx:11]: wait im dumb LMFAO

File: solutions/platinum/usaco-816.mdx (hunk 7 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx:146]: wtf i did not know that

idk, I personally think its fine the way it is

File: solutions/platinum/usaco-816.mdx (hunk 5 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-816.mdx:143]: yah

File: solutions/platinum/usaco-816.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/platinum/usaco-816.mdx:146]: idt this impacts readability too much, but i'd understand if ryan wants `llabs` as it's more canonical in a sense

**Review by ryanchou-dev — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

**Pull request merged**